cache/lib/cache_web/controllers/module_cache_controller.ex

```diff
@@ -0,0 +1,331 @@
+      {:error, conn_after} -> send_resp(conn_after, :no_content, "")
+    end
+  end
+
+  defp save_new_artifact(conn, account_handle, project_handle, category, hash, name) do
+    case BodyReader.read(conn) do
+      {:ok, data, conn_after} ->
+        size = get_data_size(data)
+        :telemetry.execute([:cache, :module, :upload, :attempt], %{size: size}, %{})
+        persist_artifact(conn_after, account_handle, project_handle, category, hash, name, data, size)
+
+      {:error, :too_large, conn_after} ->
+        :telemetry.execute([:cache, :module, :upload, :error], %{count: 1}, %{reason: :too_large})
+        send_error(conn_after, :request_entity_too_large, "Request body exceeded allowed size")
+
+      {:error, :timeout, conn_after} ->
+        :telemetry.execute([:cache, :module, :upload, :error], %{count: 1}, %{reason: :timeout})
+        send_error(conn_after, :request_timeout, "Request body read timed out")
+
+      {:error, :cancelled, conn_after} ->
+        :telemetry.execute([:cache, :module, :upload, :cancelled], %{count: 1}, %{})
+        send_resp(conn_after, :no_content, "")
+
+      {:error, _reason, conn_after} ->
+        :telemetry.execute([:cache, :module, :upload, :error], %{count: 1}, %{reason: :read_error})
+        send_error(conn_after, :internal_server_error, "Failed to persist artifact")
+    end
+  end
+
+  defp get_data_size({:file, tmp_path}) do
+    case File.stat(tmp_path) do
+      {:ok, %File.Stat{size: sz}} -> sz
+      _ -> 0
+    end
+  end
+
+  defp get_data_size(bin) when is_binary(bin), do: byte_size(bin)
+
+  defp persist_artifact(conn, account_handle, project_handle, category, hash, name, data, size) do
+    case Disk.module_put(account_handle, project_handle, category, hash, name, data) do
+      :ok ->
+        key = Disk.module_key(account_handle, project_handle, category, hash, name)
+
+        :telemetry.execute([:cache, :module, :upload, :success], %{size: size}, %{
+          category: category,
+          hash: hash,
+          name: name,
+          account_handle: account_handle,
+          project_handle: project_handle
+        })
+
+        :ok = CacheArtifacts.track_artifact_access(key)
+        S3Transfers.enqueue_module_upload(account_handle, project_handle, category, hash, name)
+        send_resp(conn, :no_content, "")
+
+      {:error, :exists} ->
+        :telemetry.execute([:cache, :module, :upload, :exists], %{count: 1}, %{})
+        send_resp(conn, :no_content, "")
+
+      {:error, _reason} ->
+        :telemetry.execute([:cache, :module, :upload, :error], %{count: 1}, %{reason: :persist_error})
+        send_error(conn, :internal_server_error, "Failed to persist artifact")
+    end
+  end
+
+  defp send_error(conn, status, message) do
+    conn
+    |> put_status(status)
+    |> json(%{message: message})
+  end
+
+  defp send_not_found(conn) do
+    conn
+    |> put_status(:not_found)
+    |> json(%{message: "Artifact not found"})
+  end
```

Comment:
What if we have an error handler for `Accept: application/json` requests to the API such that we can raise from the controller and let the handler format all the errors in the same way?